### DISCUSSION TITLE
Intelligent 'Dismiss stale approvals when new commits are pushed'

### DISCUSSION THREAD

**Discussion by sandstrom** [community/community#12876] [Pull Requests] [answered]
There is a Github branch protection setting called "Dismiss stale pull request approvals when new commits are pushed". Unfortunately, this one won't allow a workflow with fixup and squash commits.

## Feature Request

During a PR review with it's common that 1-2 fixup commits are added to address reviewer concerns. It would be good if the PR approvals were not dismissed, in the scenario where commits are squashed, but no actual files has changed, i.e. there are no code changes, only commits being moved around or merged.

## Example Scenario

1. A developer adds a PR with the following commits:

```
- Add first name to user model (a931bb)
- Update user icon (72ce49)
- Adjust user profile wording (eff2aa)
```

2. During the review phase, the author adds two fixup commits, to address concerns that came up. The commits are now:

```
- Add first name to user model (a931bb)
- Update user icon (72ce49)
- Adjust user profile wording (eff2aa)
- fixup! Add first name (1457e3)
- fixup! Adjust user profile (bb2445)
```

3. The reviewer adds approval to this PR, everything is good!

4. Now, the problem is, if the author would merge the two fixup commits, the approval would be dismissed as stale, even though there is no difference in the code, only two merged PRs. 
    
**Solution:** Would be great if a SHA-hash of the entire diff (or each individual touched file in PR) was compared instead, and if they are identical the existing approval is *not* considered stale, even if commits have changed.

***

![image](https://user-images.githubusercontent.com/122287/158359832-c02e5710-a7c5-4edf-932d-cec503f838a2.png)

**Comment by giner** [comment]
While this may seem to be a minor feature it would improve development flows a lot at low cost

**Reply by sandstrom** [reply]
Write directly to Github (their product feedback form) and let them know. They rarely read this forum.

**Comment by singhprincejeet** [comment]
Rebasing my changes with latest commit without any code change should not lead to dismissal of review either. This promotes people to push stale code v/s getting their code reviewed again.

**Reply by bewuethr** [reply]
I'm pretty sure that as of today, an "update with rebase" (at least via the UI) does not require re-review, even if "dismiss stale approvals when new commits are pushed" is enabled.

**Reply by sandstrom** [reply]
That's not what this is about. 

It's about the scenario where you have a few "fixup! My other commit" commits present when the PR was approved.

But as a last step before merging, you want to squash them in (but the actual file diff across the full PR is unchanged). 

Doing so will invalidate the review.

**Reply by bewuethr** [reply]
I know that it doesn't fix *your* issue with fixup/squash, but the rebase issue *has* been fixed.

**Reply by williamokano-dh** [reply]
But you can squash through the interface and it will work. I know it doesn't provide all the possibilities like doing through the UI, but it works relatively well. Also, squashing changes the tree, so it identifies as a new commit.

Fixing this would have to change from very easy checks (approved commit) by checking the whole contents change

**Comment by rwultsch** [comment]
This request fixes an incessant annoyance for me.

**Reply by sandstrom** [reply]
Write directly to Github (their product feedback form) and let them know. They rarely read this forum.

**Comment by jimbali** [comment]
If even Bitbucket can do it then surely GitHub can? 🤔

**Reply by nslusher-sf** [reply]
I also found this surprising as GitHub's feature set is far superior in every other aspect.

**Comment by shlomiken** [comment]
The problem i see is this , our branch is defined to be up-to-date , so if someone approved my PR and i need to rebase the approval goes away, this is very annoying as it might happen couple of times a day.

**Comment by ghost** [comment]
![image](https://user-images.githubusercontent.com/34514071/220808568-3f9f0b89-ae5e-4108-b0b6-a7f9345bd75d.png)
If you use the rebase branch button in GitHub, the approval does not get dismissed, even with 'Dismiss stale...' checked.

**Reply by sandstrom** [reply]
That's not what this issue is about. See the text under "Example Scenario" above. It's about squashing commits together, even though the overall diff for the PR remains unchanged.

**Comment by DarylCantrell** [comment, marked as answer]
@sandstrom @shlomiken

Thanks for the feedback. As part of the security changes we made to review policies (see [this blog post](https://github.blog/changelog/2023-06-06-security-enhancements-to-required-approvals-on-pull-requests/)), we also enabled the use-cases you are asking about:

- rebasing locally and pushing the rebased source branch
- merging the target into source branch without conflicts
- splitting commits apart without making code changes
- squashing commits together without making code changes

...or any combination of those operations.

Regards, Daryl

**Reply by sandstrom** [reply]
Awesome, thanks @DarylCantrell! 💯 

When I have the ear of someone on the team: another long-running problem we have with Github PR flows is the lack of support for semi-linear merge.

- https://github.com/orgs/community/discussions/8940
- https://github.com/orgs/community/discussions/14863

Both for Merge Queue and generally.

It's supported in Gitlab, Azure DevOps and other tools. More details in the two issues above.

**Reply by DarylCantrell** [reply]
Love semi-linear merges. How much? I was the one who added them to Azure DevOps some years back. 😆 

I'll ping the PR team about this, but they have their own backlog so I can't promise a thing.

**Reply by sandstrom** [reply]
Wow! Alright, sounds like you're the best advocate we can have for it! 😄

**Reply by ZacSweers** [reply]
@DarylCantrell are you saying those were added in June or are being added now in addition to the changes made in June?

**Reply by willsmythe** [reply]
These changes landed in June, but we neglected to highlight them in the June changelog post.

**Reply by dalmolinz** [reply]
A very obvious use case that's missing is when applying suggestions. I.e.:

1. A opens PR
2. B approves with suggestions
3. A applies suggestions, say directly from the GH UI 
4. The review is dismissed (!?)

**Reply by Wind010** [reply]
It would be nice to have a manifest of files that would require re-approval if changed.  Would also be nice to have some sort of configuration for Levenshtein distance threshold between files in last commit vs the latest commit.

**Reply by tm-slavik57** [reply]
@DarylCantrell 
It seems like `rebasing locally and pushing the rebased source branch` still dismisses the review.
The only option to rebase without dismissing the review is via the GitHub UI.
Am I missing something?

**Reply by goldjacobe** [reply]
@tm-slavik57 I wrote an action that handles the rebase case: https://github.com/withgraphite/dismiss-stale-approvals
Could probably adjust this slightly to handle the squash case as well, if any future searchers find this!. You'd disable the rule and use the action instead.

**Reply by davidnewhall** [reply]
The use case that @dalmolinz described is feature requested here: https://github.com/orgs/community/discussions/78039

**Comment by akloss-terion** [comment]
YMMV, but I have observed in the last few weeks that GitHub's stale review dismissal is much improved. I've tried a few sorts of rebase/squash operations and it seems if the committed text is unchanged, GitHub no longer marks the review as stale.
